### PR TITLE
fixes simulation

### PR DESCRIPTION
**Pull request by jucheval**
- add time sorting for Poisson processes
- change Inf to typemax in simulate_ogata
- add unit tests for Poisson process with null intensity

_I wanted to make two PRs but I made a mistake and they are combined into one._

### CODE REVIEW

File: src/simulation.jl [hunk 1]
```diff
@@ -10,11 +10,11 @@ function simulate_ogata(
     t = tmin
     while t < tmax
         B, L = ground_intensity_bound(pp, t + eps(t), h)
-        τ = B > 0 ? rand(rng, Exponential(inv(B))) : Inf
+        τ = B > 0 ? rand(rng, Exponential(inv(B))) : typemax(inv(B))
         if τ > L
             t = t + L
         elseif τ <= L
-            U = rand(rng, T)
+            U = rand(rng)
```
Comment (gdalle) [inline src/simulation.jl:17]:
```suggestion
            U = rand(rng, T)
```
This directly samples a variable with the right type `T` instead of `Float64`

File: src/simulation.jl [hunk 2]
```diff
@@ -10,11 +10,11 @@ function simulate_ogata(
     t = tmin
     while t < tmax
         B, L = ground_intensity_bound(pp, t + eps(t), h)
-        τ = B > 0 ? rand(rng, Exponential(inv(B))) : Inf
+        τ = B > 0 ? rand(rng, Exponential(inv(B))) : typemax(inv(B))
```
Comment (gdalle) [inline src/simulation.jl:13]:
```suggestion
        τ = B > 0 ? rand(rng, Exponential(T(inv(B)))) : typemax(T)
```
We already have the right type at our disposal: it is `T`

File: src/simulation.jl [hunk 3]
```diff
@@ -10,11 +10,11 @@ function simulate_ogata(
...
         if τ > L
             t = t + L
```
Comment (gdalle) [inline src/simulation.jl:15]:
```suggestion
            t = t + T(L)
```

File: src/poisson/simulation.jl [hunk 4]
```diff
@@ -3,7 +3,7 @@ function Base.rand(
 ) where {M,T<:Real}
     mark_dist = mark_distribution(pp)
     N = rand(rng, Poisson(float(ground_intensity(pp) * (tmax - tmin))))
-    times = rand(rng, Uniform(tmin, tmax), N)
+    times = sort(rand(rng, Uniform(tmin, tmax), N))
```
Comment (gdalle) [inline src/poisson/simulation.jl:6]: Good catch!
```suggestion
    times = sort!(rand(rng, Uniform(tmin, tmax), N))
```
This would be even more efficient because it sorts in-place

**Review by gdalle — CHANGES_REQUESTED** [review]
Hi, thank you for this first contribution! I made a few minor remarks but overall this is nearly good to go :)

**Comment by gdalle** [comment]
The tests are failing for different reasons, let me fix those for you

**Comment by jucheval** [comment]
FYI, I also looked at the tests. I added these lines to Project.toml in the compat section :

LinearAlgebra = "1.11"
Random = "1.11"
Statistics = "1.11"
Aqua = "0.8"
Documenter = "1.8"
ForwardDiff = "0.10"
JuliaFormatter = "1.0"
Test = "1.11"
Zygote = "0.6"

**Comment by jucheval** [comment]
Test throws also a warning for multiple definition of "f" in "test/marked_poisson_process.jl" and "test/multivariate_poisson_process.jl". We could rename them by "ll_markedpp" and "ll_multipp"

File: src/simulation.jl (hunk 1 above)
Comment (jucheval) [reply src/simulation.jl:17]: I am not sure that type T is relevant here. If I am correct, T is the type of the time variables. However, U is more an "intensity variable" (in particular it is compared with intensity_bound which has no reason to be of type T).

File: src/simulation.jl (hunk 2 above)
Comment (jucheval) [reply src/simulation.jl:13]: Aren't any chance that `T(inv(B))` throws an error ?
For instance, if `T == Int`... I agree that would not be the case in practice but it is at least possible.

File: src/poisson/simulation.jl (hunk 4 above)
Comment (jucheval) [reply src/poisson/simulation.jl:6]: I agree !

**Comment by gdalle** [comment]
> FYI, I also looked at the tests. I added these lines to Project.toml in the compat section :

Yeah there were missing compat bounds but a lot of other stuff is outdated so I'll take care of it all in a separate PR

File: src/simulation.jl (hunk 3 above)
Comment (jucheval) [reply src/simulation.jl:15]: I agree, but I would say that this is handled by automatic conversion of Julia, no ?
By the way, `ground_intensity_bound` should return an `L` of type `T`.

File: src/simulation.jl (hunk 2 above)
Comment (gdalle) [reply src/simulation.jl:13]: No because inv(2) returns a floating point number, it doesn't throw an error

File: src/simulation.jl (hunk 1 above)
Comment (gdalle) [reply src/simulation.jl:17]: Essentially we have to decide which number type we want to get for the times when simulating. It can either be
- the type of the temporal bounds tmin and tmax (that is the current approach, albeit imperfectly implemented)
- the type defined implicitly by the point process (for example the type of inv(lambda))
- the promotion of the two to the most generic type

It is an arbitrary design choice, we just need to pick one option. The reason I like the current behavior is because you can easily know the type ahead of time, even when you never interact with the point process.

File: src/simulation.jl (hunk 2 above)
Comment (jucheval) [reply src/simulation.jl:13]: My example is:
- assume `T=Int` and `B=2`.
- Then, `inv(2) = .5 ::Float64` and `T(inv(B))`, i.e. `Int(.5)` throws an error.

Comment (gdalle) [reply src/simulation.jl:13]: Oh right. But then the code would error anyway because we would try to create an array of times with element type `Int` and then push real numbers into it. So I would say this doesn't change much

File: src/simulation.jl (hunk 1 above)
Comment (gdalle) [reply src/simulation.jl:17]: I understand what you mean now, and you're right, `U` is unitless here

File: src/simulation.jl [hunk 5]
```diff
@@ -10,11 +10,11 @@ function simulate_ogata(
...
         elseif τ <= L
-            U = rand(rng, T)
+            U = rand(rng)
             if U < ground_intensity(pp, t + τ, h) / B
```
Comment (gdalle) [inline src/simulation.jl:18]:
```suggestion
            if U < ground_intensity(pp, t + τ, h) / T(B)
```

File: src/simulation.jl (hunk 2 above)
Comment (gdalle) [reply src/simulation.jl:13]: Actually you're right, we can keep `inv(B)` here. The conversion will happen anyway if necessary and it's more elegant

File: src/simulation.jl (hunk 5 above)
Comment (gdalle) [inline src/simulation.jl:18]:
```suggestion
            U_max = ground_intensity(pp, t + τ, h) / B
            U = rand(rng, typeof(U_max))
            if U < U_max
```

Comment (gdalle) [reply src/simulation.jl:18]: how about this?

Comment (jucheval) [reply src/simulation.jl:18]: OK

**Review by gdalle — APPROVED** [review]
congrats on your first merged PR! looking forward to many more :)
